src/main/java/io/netty/buffer/api/CompositeBuf.java

```diff
@@ -700,6 +701,27 @@ public void compact() {
         writerOffset(woff - distance);
     }
 
+    @Override
+    public int componentCount() {
+        int sum = 0;
+        for (Buf buf : bufs) {
+            sum += buf.componentCount();
+        }
+        return sum;
```

Comment:
This could in theory overflow. I think we need to guard against it.

Reply:
It could, since we don't prevent creating zero-sized buffers, or prevent composite buffers from containing themselves.

Throwing an exception at this point would be inconvenient, but returning a `long` could also overflow after a few hundred years. Not sure which approach is better.

Reply:
Hmm… I think we can still get away with checking this up front, because only owned composite buffers can extended, and the only way for someone to extend an owned composite buffer that is a constituent component of another composite buffer, is to hold a reference to it after a `close()` call. Doing that is obviously undefined. Then, we can also filter away buffers that are empty, since there's no way for them to contribute to a composite buffer.

With all that in place, we are safe against overflow because now the number of components can not be greater than the number of bytes, _except_ when a composite buffer contains itself, in which case the number of components is infinite.

Reply:
> Hmm… I think we can still get away with checking this up front, because only owned composite buffers can extended, and the only way for someone to extend an owned composite buffer that is a constituent component of another composite buffer, is to hold a reference to it after a close() call. Doing that is obviously undefined. Then, we can also filter away buffers that are empty, since there's no way for them to contribute to a composite buffer.

Hmm... How would this guard up fron if the `Buf` implementation may be custom ?

> With all that in place, we are safe against overflow because now the number of components can not be greater than the number of bytes, except when a composite buffer contains itself, in which case the number of components is infinite.

We should ensure you can not add "yourself" to a `composite buffer`

Reply:
> Hmm... How would this guard up fron if the `Buf` implementation may be custom ?

This is in the spec of the `Buf` interface, and for composite buffers. Buffers must be owned to change capacity; buffers that become part of a composite buffer have their ref counts incremented (preventing ownership if there are other references); if you relinquish your reference to a buffer, you're not allowed to touch it anymore.

How the buffers are implemented won't matter as long as they are compliant.

> We should ensure you can not add "yourself" to a `composite buffer`

We'll have to unwrap any nesting before we can make this check. I have a card for this already.